src/data_race.rs

```diff
@@ -542,6 +542,35 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: MiriEvalContextExt<'mir, 'tcx> {
         Ok(old)
     }
 
+    /// Perform an conditional atomic exchange with a memory place and a new
+    /// scalar value, the old value is returned.
+    fn atomic_min_max_scalar(
+        &mut self,
+        place: &MPlaceTy<'tcx, Tag>,
+        rhs: ImmTy<'tcx, Tag>,
+        min: bool,
+        atomic: AtomicRwOp,
+    ) -> InterpResult<'tcx, ImmTy<'tcx, Tag>> {
+        let this = self.eval_context_mut();
+
+        let old = this.allow_data_races_mut(|this| this.read_immediate(&place.into()))?;
+
+        // `binary_op` will bail if either of them is not a scalar.
+        let cond = if min {
+            this.overflowing_binary_op(mir::BinOp::Lt, &old, &rhs)?.0
+        } else {
+            this.overflowing_binary_op(mir::BinOp::Gt, &old, &rhs)?.0
+        };
```

Comment:
I think this is swapped. Afaict if it's a min operation, when `old < rhs` we should not overwrite the value, but the current code does it the other way around.

Reply:
@oli-obk What should we do instead of overwriting the value when it's a min operation?

Reply:
I think the conclusion from the discussion in https://github.com/rust-lang/miri/pull/1653 was that this should *always* overwrite the value. Just like `fetch_add(0)` is always a write access (for the purpose of data race detection), so is `fetch_max(usize::MAX)`.

Reply:
well, you still need to find out the value of `cond` to know whether to use the new or the old value to write. In either case, the branches on this `if` condition are swapped. This does a `min` operation for `max` and vice versa.

Reply:
@oli-obk Do you mean that `!cond` then we `this.allow_data_races_mut(|this| this.write_immediate_to_mplace(*rhs, place))?`?

Reply:
I think the code should be more like
```rust
let lt = this.overflowing_binary_op(mir::BinOp::Lt, &old, &rhs)?.0.to_bool()?;
let new_val = if min {
  if lt { &old } else { &rhs }
} else /* max */ {
  if lt { &rhs } else { &old }
};
this.allow_data_races_mut(|this| this.write_immediate_to_mplace(new_val, place))?;
```